README.md
# node-fcdata

FastConnect Data NodeJs client

## Getting started

To make it easy for you to get started with GitLab, here's a list of recommended next steps.

Already a pro? Just edit this README.md and make it your own. Want to make it easy? [Use the template at the bottom](#editing-this-readme)!

## Add your files

- [ ] [Create](https://docs.gitlab.com/ee/user/project/repository/web_editor.html#create-a-file) or [upload](https://docs.gitlab.com/ee/user/project/repository/web_editor.html#upload-a-file) files
- [ ] [Add files using the command line](https://docs.gitlab.com/ee/gitlab-basics/add-file.html#add-a-file-using-the-command-line) or push an existing Git repository with the following command:

```
cd existing_repo
git remote add origin https://gitlab.ssi.com.vn/developer-2-public/node-fcdata.git
git branch -M main
git push -uf origin main
```

## Integrate with your tools

- [ ] [Set up project integrations](https://gitlab.ssi.com.vn/developer-2-public/node-fcdata/-/settings/integrations)

## Collaborate with your team

- [ ] [Invite team members and collaborators](https://docs.gitlab.com/ee/user/project/members/)
- [ ] [Create a new merge request](https://docs.gitlab.com/ee/user/project/merge_requests/creating_merge_requests.html)
- [ ] [Automatically close issues from merge requests](https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
- [ ] [Enable merge request approvals](https://docs.gitlab.com/ee/user/project/merge_requests/approvals/)
- [ ] [Automatically merge when pipeline succeeds](https://docs.gitlab.com/ee/user/project/merge_requests/merge_when_pipeline_succeeds.html)

## Test and Deploy

Use the built-in continuous integration in GitLab.

- [ ] [Get started with GitLab CI/CD](https://docs.gitlab.com/ee/ci/quick_start/index.html)
- [ ] [Analyze your code for known vulnerabilities with Static Application Security Testing(SAST)](https://docs.gitlab.com/ee/user/application_security/sast/)
- [ ] [Deploy to Kubernetes, Amazon EC2, or Amazon ECS using Auto Deploy](https://docs.gitlab.com/ee/topics/autodevops/requirements.html)
- [ ] [Use pull-based deployments for improved Kubernetes management](https://docs.gitlab.com/ee/user/clusters/agent/)
- [ ] [Set up protected environments](https://docs.gitlab.com/ee/ci/environments/protected_environments.html)

***

# Editing this README

When you're ready to make this README your own, just edit this file and use the handy template below (or feel free to structure it however you want - this is just a starting point!).  Thank you to [makeareadme.com](https://www.makeareadme.com/) for this template.

## Suggestions for a good README
Every project is different, so consider which of these sections apply to yours. The sections used in the template are suggestions for most open source projects. Also keep in mind that while a README can be too long and detailed, too long is better than too short. If you think your README is too long, consider utilizing another form of documentation rather than cutting out information.

## Name
Choose a self-explaining name for your project.

## Description
Let people know what your project can do specifically. Provide context and add a link to any reference visitors might be unfamiliar with. A list of Features or a Background subsection can also be added here. If there are alternatives to your project, this is a good place to list differentiating factors.

## Badges
On some READMEs, you may see small images that convey metadata, such as whether or not all the tests are passing for the project. You can use Shields to add some to your README. Many services also have instructions for adding a badge.

## Visuals
Depending on what you are making, it can be a good idea to include screenshots or even a video (you'll frequently see GIFs rather than actual videos). Tools like ttygif can help, but check out Asciinema for a more sophisticated method.

## Installation
Within a particular ecosystem, there may be a common way of installing things, such as using Yarn, NuGet, or Homebrew. However, consider the possibility that whoever is reading your README is a novice and would like more guidance. Listing specific steps helps remove ambiguity and gets people to using your project as quickly as possible. If it only runs in a specific context like a particular programming language version or operating system or has dependencies that have to be installed manually, also add a Requirements subsection.

## Usage
Use examples liberally, and show the expected output if you can. It's helpful to have inline the smallest example of usage that you can demonstrate, while providing links to more sophisticated examples if they are too long to reasonably include in the README.

## Support
Tell people where they can go to for help. It can be any combination of an issue tracker, a chat room, an email address, etc.

## Roadmap
If you have ideas for releases in the future, it is a good idea to list them in the README.

## Contributing
State if you are open to contributions and what your requirements are for accepting them.

For people who want to make changes to your project, it's helpful to have some documentation on how to get started. Perhaps there is a script that they should run or some environment variables that they need to set. Make these steps explicit. These instructions could also be useful to your future self.

You can also document commands to lint the code or run tests. These steps help to ensure high code quality and reduce the likelihood that the changes inadvertently break something. Having instructions for running tests is especially helpful if it requires external setup, such as starting a Selenium server for testing in a browser.

## Authors and acknowledgment
Show your appreciation to those who have contributed to the project.

## License
For open source projects, say how it is licensed.

## Project status
If you have run out of energy or time for your project, put a note at the top of the README saying that development has slowed down or stopped completely. Someone may choose to fork your project or volunteer to step in as a maintainer or owner, allowing your project to keep going. You can also make an explicit request for maintainers.


ssi-fcdata/CHANGELOG.md
# FCData NodeJS Client Changelog

# [2.2.0](https://github.com/SSI-Securities-Corporation/node-fcdata/compare/v2.1.2...v2.2.0) (2023-07-27)


### Features

* update new domain for data hub streaming ([5e5b620](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/5e5b620d16b90b627006228cb4077c870937b38e))

## [2.1.2](https://github.com/SSI-Securities-Corporation/node-fcdata/compare/v2.1.1...v2.1.2) (2023-07-18)


### Bug Fixes

* update docs and example ([f9691e3](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/f9691e30c4c4f2aab2d43f6558b757c2eae3d9be))

## [2.1.1](https://github.com/SSI-Securities-Corporation/node-fcdata/compare/v2.1.0...v2.1.1) (2023-07-18)


### Bug Fixes

* update docs ([dd4a46b](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/dd4a46b731de05ce93519d00bbc458eb3bca931b))

# [2.1.0](https://github.com/SSI-Securities-Corporation/node-fcdata/compare/v2.0.0...v2.1.0) (2023-06-20)


### Features

* update zip ([822ad5e](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/822ad5e9afe55250599dddf88bb147a4c8aa730d))

# [2.0.0](https://github.com/SSI-Securities-Corporation/node-fcdata/compare/v1.0.0...v2.0.0) (2023-06-20)


### Features

* update readme ([79e0132](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/79e0132f905e03e9d44b90ec7ee51cc0f1490f64))


### BREAKING CHANGES

* remove protobuf v1 version

# 1.0.0 (2023-06-20)


### Bug Fixes

* remove ([c30f7f4](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/c30f7f4b9a12598d973d388650b8e780cf53296a))
* update gitignore ([f527090](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/f527090c24fb0a973851c058652d5792ea371120))


### Features

* nomalize api ([16d641a](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/16d641a8589ff07e697fe7e115f661a2490db386))
* update git ignore ([79fac25](https://github.com/SSI-Securities-Corporation/node-fcdata/commit/79fac2547576963222ef5350b6c9d053c753e034))


### BREAKING CHANGES

* change lib folder


ssi-fcdata/index.js

const signalr = require("./signalR");


function addSlash(str) {
    return str.substr(-1) !== "/" ? (str + "/") : str
}

var api = {
    GET_ACCESS_TOKEN: "api/v2/Market/AccessToken",
    GET_SECURITIES_LIST: "api/v2/Market/Securities",
    GET_SECURITIES_DETAILs: "api/v2/Market/SecuritiesDetails",
    GET_INDEX_COMPONENTS: "api/v2/Market/IndexComponents",
    GET_INDEX_LIST: "api/v2/Market/IndexList",
    GET_DAILY_OHLC: "api/v2/Market/DailyOhlc",
    GET_INTRADAY_OHLC: "api/v2/Market/IntradayOhlc",
    GET_DAILY_INDEX: "api/v2/Market/DailyIndex",
    GET_DAILY_STOCKPRICE: "api/v2/Market/DailyStockPrice",
    SIGNALR: "v2.0/signalr"
}

function resoleURL(baseURL, query) {
    return addSlash(baseURL) + query;
}
var constants = {
    AUTHORIZATION_HEADER: "Authorization",
    AUTHORIZATION_SCHEME: "Bearer"
}
var client = {};
var events = {
    onData: "onData",
    onError: "onError",
    onConnected: "onConnected",
    onReconnecting: "onReconnecting",
    onReconnected: "onReconnected"
}
exports.streamClient = client;
exports.api = api;
exports.constants = constants;

/**
 * Events streaming
 */
exports.events = events;
/**
 * Init client stream order
 * @param {{url: string, token: "Bearer <accessToken>"}} options
 */
exports.initStream = function (options) {
    var url = resoleURL(options.url, api.SIGNALR);
    client = new signalr.client(
        url,
        ["FcMarketDataV2Hub"],
        10,
        true
    );
    client.serviceHandlers.connected = function (connection) {
        if (client._eventsListener.hasOwnProperty(events.onConnected)) {
            client._eventsListener[events.onConnected]();
        }
    }
    client.serviceHandlers.reconnected = function (connection) {
        if (client._eventsListener.hasOwnProperty(events.onReconnected)) {
            client._eventsListener[events.onReconnected]();
        }
    }
    client.serviceHandlers.reconnecting = function (connection) {
        if (client._eventsListener.hasOwnProperty(events.onReconnecting)) {
            client._eventsListener[events.onReconnecting]();
        }
    }
    client._eventsListener = [];
    client.headers['Authorization'] = options.token;

    client.on("FcMarketDataV2Hub", "Broadcast", function (message) {
        if (client._eventsListener.hasOwnProperty(events.onData)) {
            client._eventsListener[events.onData](message);
        }
    });

    client.on("FcMarketDataV2Hub", "Reconnected", function (message) {
        console.log("Reconnected" + message);
    });

    client.on("FcMarketDataV2Hub", "Disconnected", function (message) {
        console.log("Disconnected" + message);
    });

    client.on("FcMarketDataV2Hub", "Error", function (message) {
        console.log(message);
        if (client._eventsListener.hasOwnProperty(events.onError)) {
            client._eventsListener[events.onError](events.onError, message);
        }
    });
}

/**
 * Start listen stream from server.
 */
exports.start = function start() {
    client.start();
    return client;
}
/**
 * Subcribe event from server
 * @param {string} event value of {@link exports.events}
 * @param {(data: {})=>void} func delegate
 */
exports.bind = function (event, func) {
    //eventsListener.on(event, func);
    client._eventsListener[event] = func;
}
/**
 * Un-Subcribe event from server
 * @param {string} event value of {@link exports.events}
 * @param {(data: {})=>void} func delegate
 */
exports.unbind = function (event, func) {
    //eventsListener.removeListener(event, func);
    delete client._eventsListener[event];
}
/**
 * Switch channel
 * @param {string} channel select channel to subcribe X:ALL, X-TRADE:ALL bla bla ....
 * 
 */
exports.switchChannel = function (channel) {
    //eventsListener.removeListener(event, func);
    client.invoke(
        'FcMarketDataV2Hub',
        'SwitchChannels',
        channel
      );
}

FCData_JS/Streamings/marketStreaming.js

const signalr = require("signalr-client");


function addSlash(str) {
    return str.substr(-1) !== "/" ? (str + "/") : str
}

var api = {
    SIGNALR: "signalr"
}

function resoleURL(baseURL, query) {
    return addSlash(baseURL) + query;
}

var client = {};

/**
 * Init client stream order
 * @param {{url: string, consumer_id:string,consumer_secret:string}} options
 */
exports.initStream = function (options) {
    var url = resoleURL(options.url, api.SIGNALR);
    client = new signalr.client(
        url,
        ["FcMarketDataV2Hub"],
        10,
        true
    );

    client._eventsListener = [];
    client.headers['Authorization'] = options.token;

    client.on("FcMarketDataV2Hub", "Broadcast", function (message) {
        console.log(message);
    });

    client.on("FcMarketDataV2Hub", "Reconnected", function (message) {
        console.log("Reconnected" + message);
    });

    client.on("FcMarketDataV2Hub", "Disconnected", function (message) {
        console.log("Disconnected" + message);
    });

    client.on("FcMarketDataV2Hub", "Error", function (message) {
        console.log(message);
    });
}
exports.streamClient = client;

/**
 * Start listen stream from server.
 */
exports.start = function start() {
    client.start();
    return client;
}



main.js
/** @START_CONFIG */

const express = require('express');
const config = require('./config.js');
const client = require('ssi-fcdata');
const axios = require('axios');
const app = express();
const port = 3020;
/** @END_CONFIG */

app.get('/Securities', (req, res) => {
  let lookupRequest = {}
  lookupRequest.market = "HOSE";
  lookupRequest.pageIndex = 4;
  lookupRequest.pageSize = 100;
  Object.assign(lookupRequest, req.query)

  axios.get(config.market.ApiUrl +  client.api.GET_SECURITIES_LIST
    + "?lookupRequest.market=" + lookupRequest.market
    + "&lookupRequest.pageIndex=" + lookupRequest.pageIndex
    + "&lookupRequest.pageSize=" + lookupRequest.pageSize)
    .then(response => {
      res.send(JSON.parse(JSON.stringify(response.data)));
    }).catch(error => { console.log(error); });
});

app.get('/SecuritiesDetails', (req, res) => {
  let lookupRequest = {}
  lookupRequest.market = "DER";
  lookupRequest.symbol = "";
  lookupRequest.pageIndex = 1;
  lookupRequest.pageSize = 1000;
  Object.assign(lookupRequest, req.query)
  axios.get(config.market.ApiUrl +  client.api.GET_SECURITIES_DETAILs
    + "?lookupRequest.market=" + lookupRequest.market
    + "&lookupRequest.pageIndex=" + lookupRequest.pageIndex
    + "&lookupRequest.pageSize=" + lookupRequest.pageSize
    + "&lookupRequest.symbol=" + lookupRequest.symbol)
    .then(response => {
      res.send(JSON.parse(JSON.stringify(response.data)));
    }).catch(error => { console.log(error); });
});

app.get('/IndexComponents', (req, res) => {
  let lookupRequest = {}
  lookupRequest.indexCode = "";
  lookupRequest.pageIndex = 1;
  lookupRequest.pageSize = 1000;
  Object.assign(lookupRequest, req.query)
  axios.get(config.market.ApiUrl + client.api.GET_INDEX_COMPONENTS
    + "?lookupRequest.indexCode=" + lookupRequest.indexCode
    + "&lookupRequest.pageIndex=" + lookupRequest.pageIndex
    + "&lookupRequest.pageSize=" + lookupRequest.pageSize)
    .then(response => {
      res.send(JSON.parse(JSON.stringify(response.data)));
    }).catch(error => { console.log(error); });
});

app.get('/IndexList', (req, res) => {
  let lookupRequest = {}
  lookupRequest.exchange = "HOSE";
  lookupRequest.pageIndex = 1;
  lookupRequest.pageSize = 1000;
  Object.assign(lookupRequest, req.query)
  axios.get(config.market.ApiUrl + client.api.GET_INDEX_LIST
    + "?lookupRequest.exchange=" + lookupRequest.exchange
    + "&lookupRequest.pageIndex=" + lookupRequest.pageIndex
    + "&lookupRequest.pageSize=" + lookupRequest.pageSize)
    .then(response => {
      res.send(JSON.parse(JSON.stringify(response.data)));
    }).catch(error => { console.log(error); });
});

app.get('/DailyOhlc', (req, res) => {
  let lookupRequest = {}
  lookupRequest.symbol = "VN30F2112";
  lookupRequest.fromDate = "06/12/2021";
  lookupRequest.toDate = "16/12/2021";
  lookupRequest.pageIndex = 1;
  lookupRequest.pageSize = 1000;
  lookupRequest.ascending = true;
  Object.assign(lookupRequest, req.query)
  axios.get(config.market.ApiUrl + client.api.GET_DAILY_OHLC
    + "?lookupRequest.symbol=" + lookupRequest.symbol
    + "&lookupRequest.fromDate=" + lookupRequest.fromDate
    + "&lookupRequest.toDate=" + lookupRequest.toDate
    + "&lookupRequest.pageIndex=" + lookupRequest.pageIndex
    + "&lookupRequest.pageSize=" + lookupRequest.pageSize
    + "&lookupRequest.ascending=" + lookupRequest.ascending)
    .then(response => {
      res.send(JSON.parse(JSON.stringify(response.data)));
    }).catch(error => { console.log(error); });
});

app.get('/IntradayOhlc', (req, res) => {
  let lookupRequest = {}
  lookupRequest.symbol = "VN30F1M";
  lookupRequest.fromDate = "15/11/2021";
  lookupRequest.toDate = "15/12/2021";
  lookupRequest.pageIndex = 1;
  lookupRequest.pageSize = 1000;
  lookupRequest.ascending = false;
  Object.assign(lookupRequest, req.query)
  axios.get(config.market.ApiUrl + client.api.GET_INTRADAY_OHLC
    + "?lookupRequest.symbol=" + lookupRequest.symbol
    + "&lookupRequest.fromDate=" + lookupRequest.fromDate
    + "&lookupRequest.toDate=" + lookupRequest.toDate
    + "&lookupRequest.pageIndex=" + lookupRequest.pageIndex
    + "&lookupRequest.pageSize=" + lookupRequest.pageSize
    + "&lookupRequest.ascending=" + lookupRequest.ascending)
    .then(response => {
      res.send(JSON.parse(JSON.stringify(response.data)));
    }).catch(error => { console.log(error); });
});

app.get('/DailyIndex', (req, res) => {
  let lookupRequest = {}
  lookupRequest.indexId = "HNX30";
  lookupRequest.fromDate = "27/01/2021";
  lookupRequest.toDate = "27/01/2021";
  lookupRequest.pageIndex = 1;
  lookupRequest.pageSize = 1000;
  lookupRequest.ascending = true;
  Object.assign(lookupRequest, req.query)
  axios.get(config.market.ApiUrl + client.api.GET_DAILY_INDEX
    + "?lookupRequest.indexId=" + lookupRequest.indexId
    + "&lookupRequest.fromDate=" + lookupRequest.fromDate
    + "&lookupRequest.toDate=" + lookupRequest.toDate
    + "&lookupRequest.pageIndex=" + lookupRequest.pageIndex
    + "&lookupRequest.pageSize=" + lookupRequest.pageSize
    + "&lookupRequest.ascending=" + lookupRequest.ascending)
    .then(response => {
      res.send(JSON.parse(JSON.stringify(response.data)));
    }).catch(error => { console.log(error); });
});

app.get('/DailyStockPrice', (req, res) => {
  let lookupRequest = {}
  lookupRequest.symbol = "VN30F1M";
  lookupRequest.market = "";
  lookupRequest.fromDate = "01/12/2021";
  lookupRequest.toDate = "04/12/2021";
  lookupRequest.pageIndex = 1;
  lookupRequest.pageSize = 1000;
  Object.assign(lookupRequest, req.query)
  axios.get(config.market.ApiUrl + client.api.GET_DAILY_STOCKPRICE
    + "?lookupRequest.symbol=" + lookupRequest.symbol
    + "&lookupRequest.fromDate=" + lookupRequest.fromDate
    + "&lookupRequest.toDate=" + lookupRequest.toDate
    + "&lookupRequest.pageIndex=" + lookupRequest.pageIndex
    + "&lookupRequest.pageSize=" + lookupRequest.pageSize
    + "&lookupRequest.market=" + lookupRequest.market)
    .then(response => {
      res.send(JSON.parse(JSON.stringify(response.data)));
    }).catch(error => { console.log(error); });
});

const rq = axios.create({
  baseURL: config.market.ApiUrl,
  timeout: 5000
})

rq({
  url: config.market.ApiUrl + client.api.GET_ACCESS_TOKEN,
  method: 'post',
  data: {
    consumerID: config.market.ConsumerId,
    consumerSecret: config.market.ConsumerSecret,
  }
}).then(response => {
  if (response.data.status === 200) {
    let token = "Bearer " + response.data.data.accessToken;
    axios.interceptors.request.use(function (axios_config) {
      axios_config.headers.Authorization =  token;
      return axios_config;
    });

    client.initStream({
      url: config.market.HubUrl,
      token: token,
    });
    client.bind(client.events.onData, function(message){
      console.log(message)
    })
    client.bind(client.events.onConnected, function(){
      client.switchChannel("X-QUOTE:ALL")
    })
    client.start();
  } else {
    console.log(response.data.message)
  }

}, reason => {
  console.log(reason);
})

app.listen(port, 'localhost', () => console.log(`Example app listening on port ${port}!`));
